xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,455 @@
+    // number of hashes on the the ring. Other hosts might not end up with whole numbers, and
+    // that's fine (the ring-building algorithm can handle this). This preserves the original
+    // implementation's behavior: when weights aren't provided, all hosts should get an equal
+    // number of hashes. In the case where this number exceeds the max_ring_size, it's scaled
+    // back down to fit.
+    double scale = Math.min(
+        Math.ceil(normalizedMinWeight * config.minRingSize) / normalizedMinWeight,
+        (double) config.maxRingSize);
+    ring = buildRing(serverWeights, totalWeight, scale);
+
+    // Shut down subchannels for delisted addresses.
+    List<Subchannel> removedSubchannels = new ArrayList<>();
+    for (EquivalentAddressGroup addr : removedAddrs) {
+      removedSubchannels.add(subchannels.remove(addr));
+    }
+
+    // Update the picker before shutting down the subchannels, to reduce the chance of race
+    // between picking a subchannel and shutting it down.
+    updateBalancingState();
+    for (Subchannel subchann : removedSubchannels) {
+      shutdownSubchannel(subchann);
+    }
+  }
+
+  private static List<RingEntry> buildRing(
+      Map<EquivalentAddressGroup, Long> serverWeights, long totalWeight, double scale) {
+    List<RingEntry> ring = new ArrayList<>();
+    double currentHashes = 0.0;
+    double targetHashes = 0.0;
+    for (Map.Entry<EquivalentAddressGroup, Long> entry : serverWeights.entrySet()) {
+      EquivalentAddressGroup addrKey = entry.getKey();
+      double normalizedWeight = (double) entry.getValue() / totalWeight;
+      // TODO(chengyuanzhang): is using the list of socket address correct?
+      StringBuilder sb = new StringBuilder(addrKey.getAddresses().toString());
+      sb.append('_');
+      targetHashes += scale * normalizedWeight;
+      long i = 0L;
+      while (currentHashes < targetHashes) {
+        sb.append(i);
+        long hash = hashFunc.hashAsciiString(sb.toString());
+        ring.add(new RingEntry(hash, addrKey));
+        i++;
+        currentHashes++;
+        sb.deleteCharAt(sb.length() - 1);
+      }
+    }
+    Collections.sort(ring);
+    return Collections.unmodifiableList(ring);
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    if (currentState != READY) {
+      helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    for (Subchannel subchannel : subchannels.values()) {
+      shutdownSubchannel(subchannel);
+    }
+  }
+
+  private void updateBalancingState() {
+    checkState(!subchannels.isEmpty(), "no subchannel has been created");
+    ConnectivityState overallState = aggregateState(subchannels.values());
+    RingHashPicker picker = new RingHashPicker(syncContext, ring, subchannels);
+    // TODO(chengyuanzhang): avoid unnecessary reprocess caused by duplicated server addr updates
```

Comment:
How do duplicate server addr updates come into play? It looks like we only create one subchannel for duplicate addresses.

Reply:
I mean receiving another `handleResolvedAddresses()` call with exactly the same addresses and weights. That would trigger another Channel reprocess. Round robin has a check to avoid such case: https://github.com/grpc/grpc-java/blob/1a3b02b44bdd6d0422d9f5225a3f9919873a9ce0/core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java#L205

It's a small thing, probably doesn't matter though, IMO.

Reply:
Ah, I was thinking an addr was duplicated not a duplicated call to `handleResolvedAddresses()`. Makes sense. Yeah, it is a fine TODO.